### PR TITLE
[TD-99] announce changes post_commit instead of post_save

### PR DESCRIPTION
**Pull request by lukebryant**
[TD-99]
I had to add `@pytest.mark.django_db(transaction=True)` to the tests,
because otherwise they each run in their own transaction, so it's
impossible to test `on_commit` hooks. This has the unfortunate
side-effect of removing the implicit database cleanup. To deal with
this, I've had to change some of the existing tests so that they don't
make the assumption that the first object they create will have `id=1`.
We could also change the suite so that we completely clear the database
between test runs, but we'd have to re-run all the migrations after
every test.

[TD-99]: https://lantum.atlassian.net/browse/TD-99

### CODE REVIEW

File: tests/test_postsaveactions.py [hunk 1]
```diff
@@ -1,6 +1,7 @@
 import pytest
 
 from .models import (TestModel)
+from django.db import transaction
```
Comment (Phil-Barber) [inline tests/test_postsaveactions.py]: unused?
```suggestion
```

**Comment by lukebryant** [comment]
> Would this mean that any changes currently not taking place inside a transaction would not announce changes?
> Is there any way to do like:
> 
> ```
> if transaction.is_open: 
>     transaction.onComit()
> else: 
>     onSave()
> ```

No, sorry. I should have documented this somehow. 

`on_commit` runs the handler straight away if it's outside a transaction. The existing tests verify that this is the behaviour, but it might be worth making it explicit that we're testing the behaviour outside a transaction.
https://docs.djangoproject.com/en/3.0/topics/db/transactions/#django.db.transaction.on_commit

File: tests/test_postsaveactions.py (hunk 1 above)
Comment (lukebryant) [reply tests/test_postsaveactions.py]: ya, ty. Just fixed in commit amend

**Comment by lukebryant** [comment]
I chose the way I've done this using the existing post-save handlers mainly because it's easier to implement. 

The messages are still constructed post-save, but not announced until a commit. This means that the timestamps for changes are for when they're saved, not committed. I prefer it this way, but you could make a case for the timestamps being the commit time. 

Two straight downsides of this approach are:
- Transactions take longer because processing happens inside them.
- Processing time is wasted if there's a rollback, because the messages won't be sent (I think this is pretty trivial because of how rare rollbacks should be).

**Comment by Phil-Barber** [comment]
> `on_commit` runs the handler straight away if it's outside a transaction. The existing tests verify that this is the behaviour, but it might be worth making it explicit that we're testing the behaviour outside a transaction.
 
perhaps add a docstring to the class:
```
class AnnounceChangesMixin(...):
    '''
     messages build on save / transaction yada yada yada 
    ''''
```

**Comment by Phil-Barber** [comment]
> I chose the way I've done this using the existing post-save handlers mainly because it's easier to implement.
> 
> The messages are still constructed post-save, but not announced until a commit. This means that the timestamps for changes are for when they're saved, not committed. I prefer it this way, but you could make a case for the timestamps being the commit time.
> 
> Two straight downsides of this approach are:
> 
> * Transactions take longer because processing happens inside them.
> * Processing time is wasted if there's a rollback, because the messages won't be sent (I think this is pretty trivial because of how rare rollbacks should be).

I agree with this approach 👍

**Comment by iponce42** [comment]
We'll have to be even more careful when encapsulating big chunks of code with transactions, but I think it's a fair price to pay.

Good stuff :clap:  @lukebryant

**Comment by lukebryant** [comment]
Sorry about force push, I just changed the commit message for the first commit, so looking at the latter 3 will show code changes since approval.

The test coverage should be complete now for new functionality. We need to be very careful with this library, because it's used by everything, and it doesn't have release versions. If anyone can think of any more tests that might be useful, I'd be more than happy to write them.

**Comment by lukebryant** [comment]
> We'll have to be even more careful when encapsulating big chunks of code with transactions, but I think it's a fair price to pay (hint: we shouldn't be wrapping big pieces of code in transactions!).
> 
> Good stuff 👏 @lukebryant

Thanks :)

Me and nacho discussed this, but for anyone else's benefit, these changes should speed up transactions if anything, because the announces will now be made only after the transaction finishes. I have no idea what the scale of this effect is, though.

**Comment by lukebryant** [comment]
Something that's come up in QA is that this makes the order of the febe history less reliable. When we commit a load of announce_changes messages are shot off at once, and it seems like the time in the febe history is based on when the item is processed by audit, as opposed to the timestamp on the `model_changed` message. We might want to change audit to fix this behaviour, as it seems like it's an existing bug with that service that has just been made more obvious by this change.

**Comment by Phil-Barber** [comment]
This sounds like an acceptable hit to take. In the most extreme cases do you have any idea roughly how inaccurate the times would be? ie. several seconds out vs minutes vs hours

**Comment by lukebryant** [comment]
> This sounds like an acceptable hit to take. In the most extreme cases do you have any idea roughly how inaccurate the times would be? ie. several seconds out vs minutes vs hours

The error that this change adds is proportional to the length of the transaction. If we have transactions that are lasting more than a few seconds then we have bigger problems than the febe history being slightly out. I think the order of operations in febe history is a bigger concern than the absolute error in time.

**Pull request merged**